### PR TITLE
feat(controller): don't hardcode experiment ports; always create service

### PR DESCRIPTION
**Pull request by alexef**
superseeds: #2357 

fixes: #2233
Checklist:

* [ ] Either (a) I've created an [enhancement proposal](https://github.com/argoproj/argo-rollouts/issues/new/choose) and discussed it with the community, (b) this is a bug fix, or (c) this is a chore.
* [ ] The title of the PR is (a) [conventional](https://www.conventionalcommits.org/en/v1.0.0/) with a list of types and scopes found [here](https://github.com/argoproj/argo-rollouts/blob/master/.github/workflows/pr-title-check.yml), (b) states what changed, and (c) suffixes the related issues number. E.g. `"fix(controller): Updates such and such. Fixes #1234"`.  
* [x] I've signed my commits with [DCO](https://github.com/argoproj/argoproj)
* [ ] I have written unit and/or e2e tests for my change. PRs without these are unlikely to be merged.
* [ ] My builds are green. Try syncing with master if they are not. 
* [ ] My organization is added to [USERS.md](https://github.com/argoproj/argo-rollouts/blob/master/USERS.md).

### CODE REVIEW

File: docs/features/experiment.md [hunk 1]
```diff
@@ -243,5 +245,9 @@ to `experiment-baseline`, leaving the remaining 90% of traffic to the old stack.
 
 !!! note
     When a weighted experiment step with traffic routing is used, a
-    service is auto-created for each experiment template. The traffic routers use
-    this service to send traffic to the experiment pods.
\ No newline at end of file
+    Service is auto-created for each experiment template. The traffic routers use
+    this service to send traffic to the experiment pods.
+
+By default, the generated Service has the name of the ReplicaSet and inherits
```
Comment (alexef) [inline docs/features/experiment.md:251]: needs this https://github.com/argoproj/argo-rollouts/pull/2396/files to be merged first

File: experiments/service.go [hunk 2]
```diff
@@ -74,11 +73,7 @@ func (ec *experimentContext) CreateService(serviceName string, template v1alpha1
 		},
 		Spec: corev1.ServiceSpec{
 			Selector: selector,
-			Ports: []corev1.ServicePort{{
```
Comment (alexef) [inline experiments/service.go:77]: this was #2233

File: docs/features/experiment.md [hunk 3]
```diff
@@ -6,6 +6,8 @@ The Experiment CRD allows users to have ephemeral runs of one or more ReplicaSet
 running ephemeral ReplicaSets, the Experiment CRD can launch AnalysisRuns alongside the ReplicaSets.
 Generally, those AnalysisRun is used to confirm that new ReplicaSets are running as expected.
 
+A Service routing traffic to the Experiment ReplicaSet is also generated.
```
Comment (alexef) [inline docs/features/experiment.md:9]: FIXME: this doesn't hold true

File: docs/features/experiment.md (hunk 1 above)
Comment (zachaller) [reply docs/features/experiment.md:251]: has been merged

**Comment by zachaller** [comment]
In regards to the e2e tests, can you confirm for me that if we modify the code to not create the service if there is no ports defined on the pod spec. I think that should fix all the e2e tests and not require a port on the container which I think we should support

**Comment by alexef** [comment]
>  if we modify the code to not create the service if there is no ports defined on the pod spec

correct, I forgot about that. let me do this change

**Review by zachaller — APPROVED** [review]

File: rollout/experiment.go [hunk 4]
```diff
@@ -64,9 +64,7 @@ func GetExperimentFromTemplate(r *v1alpha1.Rollout, stableRS, newRS *appsv1.Repl
 			Name:     templateStep.Name,
 			Replicas: templateStep.Replicas,
 		}
-		if templateStep.Weight != nil {
```
Comment (alexef) [inline rollout/experiment.go:67]: Feels like we can use weight: 0 to keep creating the service, if automated creation is rolled back

**Pull request merged**